compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1005,16 +1005,20 @@ private boolean isXMLTypeAssignable(BType sourceType, BType targetType, Set<Type
                     }
                     return isAssignable(source, target.constraint, unresolvedTypes);
                 }
-                return isAssignable(source.constraint, target.constraint, unresolvedTypes);
+                return isAssignable(source.constraint, target, unresolvedTypes);
             }
             return true;
         }
         if (sourceTag == TypeTags.XML) {
             BXMLType source = (BXMLType) sourceType;
             if (targetTag == TypeTags.XML_TEXT) {
                 if (source.constraint != null) {
-                    return source.constraint.tag == TypeTags.NEVER ||
-                            source.constraint.tag == TypeTags.XML_TEXT;
+                    if (source.constraint.tag == TypeTags.NEVER ||
+                            source.constraint.tag == TypeTags.XML_TEXT) {
+                        return true;
+                    } else {
+                        return isAssignable(targetType, source.constraint, unresolvedTypes);
```

Comment:
public function main() {
    xml e = xml `<e/>`;
    xml:Text t = e; // this should not be allowed
}

Reply:
Still not right, consider following code.

```ballerina
public function main() {
    xml<xml> e = xml `<elem/>`;
    xml:Text t = e; // this should not be allowed 
}
```

Reply:
Please add negative test cases too